light-client/src/supervisor.rs

```diff
@@ -0,0 +1,302 @@
+                Ok(light_block) => {
+                    // SAFETY: There must be a latest trusted state otherwise verification would have failed.
+                    let trusted_state = primary
+                        .state
+                        .light_store
+                        .latest(VerifiedStatus::Verified)
+                        .unwrap();
+
+                    let outcome = self.detect_forks(&light_block, &trusted_state);
+
+                    match outcome {
+                        Some(forks) => {
+                            let mut forked = Vec::with_capacity(forks.len());
+
+                            for fork in forks {
+                                match fork {
+                                    Fork::Forked(block) => {
+                                        self.report_evidence(&block);
+                                        forked.push(block.provider);
+                                    }
+                                    Fork::Faulty(block) => {
+                                        self.peers.remove_secondary(&block.provider);
+                                    }
+                                }
+                            }
+
+                            if !forked.is_empty() {
+                                // Fork detected, exiting
+                                bail!(ErrorKind::ForkDetected(forked))
+                            }
+                        }
+                        None => {
+                            // No fork detected, exiting
+                            // TODO: Send to relayer, maybe the run method does this?
+                            return Ok(light_block);
+                        }
+                    }
+                }
+                // Verification failed
+                Err(_err) => {
+                    // Swap primary, and continue with new primary, if any
+                    self.peers.swap_primary()?;
+                    continue;
+                }
+            }
+        }
+
+        bail!(ErrorKind::NoValidPeerLeft)
+    }
+
+    fn report_evidence(&mut self, _light_block: &LightBlock) {
+        ()
+    }
+
+    #[pre(self.peers.primary().is_some())]
+    fn detect_forks(
+        &mut self,
+        light_block: &LightBlock,
+        trusted_state: &LightBlock,
+    ) -> Option<Vec<Fork>> {
+        if self.peers.secondaries().is_empty() {
+            return None;
+        }
+
+        let fork_detector = ProdForkDetector::new(); // TODO: Should be injectable
+        let result =
+            fork_detector.detect_forks(light_block, &trusted_state, self.peers.secondaries());
+
+        match result {
+            ForkDetection::Detected(forks) => Some(forks),
+            ForkDetection::NotDetected => None,
+        }
+    }
+
+    pub fn handler(&mut self) -> Handler {
+        Handler::new(self.sender.clone())
+    }
+
+    // Consume the instance here but return a runtime which will allow interaction
```

Comment:
IMO it would help immensely, if there was a simple test / example that shows how to actually interact with the instance. e.g. maybe a call to `verify_to_target_async` with a custom callback?

Reply:
Agreed. I think we are porting over the unit tests now (as part of #280) but created a specific issue for that kind of test (#316). The use case that motivated (but may not be necessary) Was for the relayer which wants to accumulate events from its subscription while it waits for the light client. Here is the pseudo-ish code which we used to orient the discussion. 

```rust
// Snippet from  Relayer event loop
let light_client = ... // some light client handler
let (sender, receiver) = ???;

while Some(event) = recever.recv() {
    match event {
        // Assume this is from subscription
        Event::NewHeader(header) => {
            let callback = sender.clone();
            light_client.validate_async(header, move |result| {
                match result {
                    Ok(header) => {
                        callback.send(Event::ValidatedHeader(header)))
                    },
                    Err(err) => {
                        callback.send(Event::FailedValidation(err));
                    }
                }
            });
        },
        Event::ValidatedHeader(header) => {
           ...
        }
        Event::FailedValidation(err) => {
            // Shut down the relayer
        }
    }
}
```

Reply:
See https://github.com/informalsystems/tendermint-rs/pull/302/commits/76857496d8be96344b1348d25b177d545715b4c7.